src/dmd/root/aav.d

```diff
@@ -121,6 +127,93 @@ extern (C++) Value dmd_aaGetRvalue(AA* aa, Key key)
     return null; // not found
 }
 
+debug
+{
+    /**
+    Returns a range of key/values for `aa`.
+
+    Returns: a range of key/values for `aa`.
+    */
+    @property auto asRange(AA* aa)
+    {
+        return AARange(aa);
+    }
+
+    private struct AARange
+    {
+        AA* aa;
+        // current index into bucket array `aa.b`
+        size_t bIndex;
+        aaA* current;
+
+        this(AA* aa)
+        {
+            if (aa)
+            {
+                this.aa = aa;
+                toNext();
+            }
+        }
+
+        @property bool empty() const { return current is null; }
+
+        @property auto front() { return current.keyValue; }
+
+        void popFront()
+        {
+            if (current.next)
+                current = current.next;
+            else
+            {
+                bIndex++;
+                toNext();
+            }
+        }
+
+        private void toNext()
+        {
+            for (; bIndex < aa.b_length; bIndex++)
+            {
+                auto next = aa.b[bIndex];
+                if (next)
```

Comment:
> One line: if (auto next = ...)
> actually that logic doesn't work...that assignment can't be done until after the loop condition is satisfied.

I don't follow you here - is this a remark on previous code?

Reply:
Woops, I thought you were talking about putting `auto next = ...` in the for loop...will change